Cereal/CerealSerialization.swift

```diff
@@ -0,0 +1,434 @@
+    case SubTree = 11
+    case IdentifyingTree = 12
+}
+
+// MARK: Helpers
+
+/// function takes a value of type `Type` and returns a byte array representation
+private func toByteArray<Type>(value: Type) -> [UInt8] {
+    var unsafeValue = value
+    return withUnsafePointer(&unsafeValue) {
+        Array(UnsafeBufferPointer(start: UnsafePointer<UInt8>($0), count: sizeof(Type)))
+    }
+}
+
+/// function takes an array of bytes and returns a value of type `Type`
+private func fromByteArray<Type>(value: [UInt8], _: Type.Type) -> Type {
+    return value.withUnsafeBufferPointer {
+        return UnsafePointer<Type>($0.baseAddress).memory
+    }
+}
+
+/// function reads an array of bytes and returns an integer value (Int, Int64 etc)
+private func bytesToInteger<Type where Type: IntegerType>(from: [UInt8]) -> Type {
+    return from.withUnsafeBufferPointer({
+        UnsafePointer<Type>($0.baseAddress).memory
+    })
+}
+
+/// function returns a number of .StringValue entries inside an array of CoderTreeValue (including subtrees)
+private func stringEntriesCount(array: [CoderTreeValue]) -> Int {
+    var result = 0
+    for item in array {
+        result += item.numberOfStringEntries()
+    }
+    return result
+}
+
+/// function returns a byte array in format [length_in_bytes] + [number of array items] + [bytes of CoderTreeValue array]
+/// this allows a decoder to set a capacity for an array of CoderTreeValue beign decoded from bytes
+private func countCapacityBytes(array: [CoderTreeValue]) -> [UInt8] {
+    var stringMap = Dictionary<String, [UInt8]>(minimumCapacity: stringEntriesCount(array))
+
+    return countCapacityBytes(array, stringMap: &stringMap)
+}
+
+/// function returns a byte array in format [length_in_bytes] + [number of array items] + [bytes of CoderTreeValue array]
+/// this allows a decoder to set a capacity for an array of CoderTreeValue beign decoded from bytes
+/// - note: stringMap dictionary is used to improve string encoding speed, see `String.writeToBuffer`
+private func countCapacityBytes(array: [CoderTreeValue], inout stringMap: [String: [UInt8]]) -> [UInt8] {
+    var result = [UInt8]()
+
+    for item in array {
+        item.writeToBuffer(&result, stringMap: &stringMap)
+    }
+
+    return toByteArray(result.count) + toByteArray(array.count) + result
+}
+
+// MARK: encoding extensions
+
+private extension String {
+    func writeToBuffer(inout buffer: [UInt8], inout stringMap: [String: [UInt8]]) {
+        // if the string was already decoded, getting from a dictionary is a cheaper operation than getting utf8 view
+        if let bytes = stringMap[self] {
+            buffer.appendContentsOf(bytes)
+            return
+        }
+
+        var array = Array(self.utf8)
+        array.insertContentsOf(toByteArray(array.count), at: 0)
+        buffer.appendContentsOf(array)
+        stringMap.updateValue(array, forKey: self)
+    }
+}
+
+private extension CoderTreeValue {
+    func numberOfStringEntries() -> Int {
```

Comment:
I don't think this will be semantically correct in that case. Computed vars expected to be a lightweight operation, which may not be true here, since we calculate this value recursively on a possibly large tree. How do you think?